### ISSUE TITLE
Compile Error with GCC 4.8.2

### ISSUE THREAD

**Issue by anhi** [BALL-Project/ball#517]
**Reported by strunk on 23 Sep 43889757 00:20 UTC**
I got a few compile errors, which I suspect to be 'actual' bugs in  Ball version 1.4.2.

Among others, this is the error:
```
PDBFileDetails.C:902:72: error: no matching function for call to BALL::NamedProperty::NamedProperty(const char [12], boost::shared_ptr<BALL::CrystalInfo>&)
    current_protein_->setProperty(NamedProperty("CRYSTALINFO", temp_ptr ));
```

temp_ptr is of basetype:
```
boost::shared_ptr<BALL::PersistentObject>   
```
                                                          
In other words:
NamedProperty("CRYSTALINFO", temp_ptr ) is not 'found'. 
temp_ptr is correctly set 
```
    if (!(current_protein_->hasProperty("CRYSTALINFO")))
    {
      boost::shared_ptr<CrystalInfo> temp_ptr(new CrystalInfo());
      current_protein_->setProperty(NamedProperty("CRYSTALINFO", temp_ptr ));
    }
```

I guess, the reason is simple: temp_ptr is a temporary and the NamedProperty constructor expects a shared_ptr reference and not a copy. Non-const references to temporaries are not allowed. The outcome is more severe however: Being a shared pointer, the reference count upon passing as reference is not increased. If this code was compiled, the shared pointer would fall out of scope and the object would be destroyed.

The solution is very simple. As ownership of the shared_ptr is actually transfered to the NamedProperty (as long as the property exists, the object has to), the NamedProperty constructor should actually get a copy of the shared_ptr instead of a reference.
If this is desired, I would start creating a patch.

Best,
Timo

**Comment by anhi** [comment]
**Commented by chahuistle on 5 Jul 44028579 21:26 UTC**
right now, only g++ 4.2 is supported, support for g++ 4.8 would be an enhancement, not a defect

**Comment by anhi** [comment]
**Commented by strunk on 27 Feb 44028630 09:14 UTC**
It is a defect. It just shows up as a compile error in g++ 4.8. The problem is already there in g++ 4.2 in my opinion. I'd consider it a bug therefore. It's just not found by g++ 4.2. The title of the bug report could be better, I agree.

Again: temp_ptr is a non-const temporary object, which is passed by reference to setProperty. It doesn't matter that it is a shared_ptr, because we are not talking about the object pointed to, but rather about the shared_ptr object itself. Once 
```
 boost::shared_ptr<CrystalInfo> temp_ptr(new CrystalInfo());
```
runs out of scope, the shared_ptr object will cease to exist. Stored references to it will become invalid.

Now NamedProperty just calls:
data_ = po; (where po is the non-const reference to the object and data_ is boost::any).
Therefore an object copy is carried out and this bug _should_ not have severe implications. But the shared_ptr might not work as intended and you will have a memory bug on your hands, if casting is involved, because the reference count will be wrong.
See, first answer here:
http://stackoverflow.com/questions/13826897/why-not-non-const-reference-to-temporary-objects

Again, the change is very trivial, because the shared_ptr only needs to be passed by value, as is always suggested using shared_ptrs (first answer here):
http://stackoverflow.com/questions/327573/c-passing-references-to-boostshared-ptr

**Comment by tkemmer** [comment]
Should be fixed with e7cc50705bed20e160bd4b637b23f57270f580a8.
